### PR TITLE
Bugfix/567 heatmap race condition

### PR DESCRIPTION
**Pull request by davidzwa**
- [x] Add heatmap tests + repro 
- [x] Add farm info service
- [x] Fix startup index/app-core race condition **OR** patch the heatmap on the spot
- [x] Changelog

Unrelated but small changes:
- [x] Generic info to debug log changes
- [x] `undefined` to "undefined" change (although not the cause of problems thusfar)
- [x] Revert `filamentManager` renaming `filamentManagerEnabled` to avoid future headaches (either rename it all, or leave it untouched)

### CODE REVIEW

File: server_src/settings/serverSettings.js [hunk 1]
```diff
@@ -108,5 +108,5 @@ class ServerSettings {
 
 module.exports = {
   ServerSettings,
-  filamentManagerEnabled: filamentManager
+  filamentManager
```
Comment (davidzwa) [inline server_src/settings/serverSettings.js:111]: Me reverting my half-assed attempt at renaming a setting. I only foresee issues if this is not done all the way thru the UI, database etc. Thats why I reverted it.

File: server_src/lib/tests/printer-clean.test.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      farmStart: new Date(),
+      heatMap: [
+        {
+          name: "Completed",
+          data: []
+        }
+      ]
+    }])
+
+    expect(PrinterClean.initFarmInformation()).rejects.toThrow("Cannot read property 'x' of undefined");
```
[the remark this answers is not in the document]
Comment (NotExpectedYet) [reply server_src/lib/tests/printer-clean.test.js:51]: We could define the heat map in the database no? We'd only have to check for the day in the code if I remember rightly.

File: server_src/settings/serverSettings.js (hunk 1 above)
Comment (NotExpectedYet) [reply server_src/settings/serverSettings.js:111]: Yeah that's a lot of places xD

File: server_src/lib/dataFunctions/printerClean.js [hunk 3]
```diff
@@ -1548,48 +1548,21 @@ class PrinterClean {
...
-      newfarmStats.save();
+
+      farmStats[0] = await FarmStatisticsService.create(farmStart, heatMap);
     } else if (typeof farmStats[0].heatMap === "undefined") {
       farmStats[0].heatMap = heatMap;
       dashboardStatistics.utilisationGraph = heatMap;
       farmStats[0].markModified("heatMap");
-      farmStats[0].save();
-    } else {
-      heatMap = farmStats[0].heatMap;
```
Comment (davidzwa) [inline server_src/lib/dataFunctions/printerClean.js:1566]: Deleted it @NotExpectedYet as it is duplicate, up to you to tell if this is creating problems down the line.
At least my test didnt detect a throw anymore, so the bug is gone.

File: server_src/lib/tests/printer-clean.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidzwa) [reply server_src/lib/tests/printer-clean.test.js]: See my change below, I think you're right but I saw incorrect duplication so maybe we can prevent this bug differently.

File: server_src/lib/dataFunctions/printerClean.js (hunk 3 above)
Comment (NotExpectedYet) [reply server_src/lib/dataFunctions/printerClean.js:1566]: I think your right here. That extra save was probably an early attempt to fix another issue.

Comment (davidzwa) [reply server_src/lib/dataFunctions/printerClean.js:1566]: Lol that makes me feel really uneasy

Comment (NotExpectedYet) [reply server_src/lib/dataFunctions/printerClean.js:1566]: I plan too, I wasn't meaning I wasn't going to haha

**Comment by NotExpectedYet** [comment]
I'll get this checked tomorrow and go through the file/job clean one once merged.

**Review by NotExpectedYet — APPROVED** [review]
Tested and all seems working from this end. 

- [x] Existing Mongo Local on PC issue was discovered: Result - Original database recovered. 
- [x] New Mongo Local on PC issue was discovered: Result - New database generated 
- [x] Existing mongo db on Atlas: Result - Original database recovered
- [x] Remote mongo db on Atlas: Result - New Database Generated

Good spot!

**Pull request merged**